[IV-24-25] Objetivo2

mmnuria opened the pull request:
Sobre la estructura del repositorio
* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas específicas para el lenguaje con el que se está trabajando?

Sobre el planteamiento
* [x] ¿Se han planteado una serie de issues, asignados a las historias deusuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

Sobre el análisis del problema
* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor, una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo objeto valor que no dependa de ningún otro (y que sea la base de muchos otros)?.

Sobre el código
* [x] ¿Todos los mensajes de commit explican el cambio, y no se limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto mínimamente viable? ¿O tengo que solicitar al product manager que precise de qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente correcto?

mmnuria committed:
Soluciona #7 elección lenguaje programación Go

mmnuria committed:
Soluciona #7 versión de Go a usar

mmnuria committed:
Soluciona #8 crea un struct para diferenciar cada tipo de turnos

mmnuria committed:
Soluciona #9 usando un struct guarda la información que contiene cada turno

mmnuria commented:
@JJ solicito revisión. Mi compañero ya me lo ha aprobado.

JJ commented:
@abbonno una "revisión" no es una "aprobación". En general, cuando me encuentro que la persona que gestiona el repo aprueba directamente la implicación es casi siempre "más trabajo para mi". Entre otras cosas, hay una [lista de comprobación para el propietario del repositorio](http://jj.github.io/IV/documentos/proyecto/2.Modelo#lista-de-comprobaci%C3%B3n-para-el-propietarioa-del-repositorio). Te aconsejaría que  la siguieras, poniéndolo en un comentario, y actuaras en consecuencia.

abbonno commented:
Es cierto, lo olvidé, lo escribo a continuación.

abbonno reviewed (commented):
Deberías crear los issues relativos a los cambios que se han realizado para cumplir con los aspectos necesarios de la historia de usuario

abbonno reviewed (commented):
Además deberías renombrar la carpeta de código según las buenas prácticas respectivas a la esctructura de un proyecto en go.

mmnuria committed:
resolución estructura buenas prácticas

mmnuria committed:
Soluciona #11 se necesita saber el contrato que tiene cada empleado para asignarle un turno

mmnuria committed:
Soluciona #12 struct y declaracion de las areas que hay en el hospital para asignar a los empleados en su turno de trabajo

mmnuria committed:
Soluciona #13 struct de la información necesaria de los empleados para asignarle un turno

mmnuria committed:
Soluciona #14 asignación y cronograma de cada empleado con su turno

mmnuria commented:
@abbonno ya he seguido tus consejos de renombrar la carpeta para las buenas prácticas. Por otro lado, he añadido issues nuevos y arreglado los commit para que se identifique cada uno con su correspondiente issue. Por mi parte creo que ya está todo correctamente, si ves algo más hazmelo saber, si no podemos pasar a la revisión con el profesor. Muchas gracias!

abbonno commented:
Perfecto, vuelvo a pasar la lista de comprobación:


* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y
evidencian comprensión de las historias de usuario? Si no es así, es mejor que
se le transmita antes de que empiece a crear cualquier código erróneo.
- Los issues existentes son correctos y suficientes.
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de
      comprobación?
- Sí, la cumple.
* [x] ¿El código creado es una modelización válida de las historias de usuario
      creadas por ti, y te sirve como base para programar la lógica de negocio?
- Sí.
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de
      programación que hayáis consensuado, tanto en identificadores, como
      nombres de ficheros, como disposición de los diferentes ficheros en el
      repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y
      aportan código para resolver el problema planteado en el mismo?
- Los ficheros siguen las buenas prácticas y los commits son informativos y refieren a los issues correctamente.
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo
ha hecho y documentado en el repositorio?
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone
correspondiente, que será el primero?
* [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en
el objetivo 4 para poder programar la lógica de negocio?
- Sí.
* [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar
directamente en el siguiente milestone, que incluye la lógica de negocio?
- Sí.

mmnuria commented:
@JJ solicito revision

JJ commented on internal/areas.go:
```diff
@@ -0,0 +1,18 @@
+	EnfermerosNecesarios int
+}
+
+var (
```
Dudo muchísimo que esto sea necesario en un producto mínimamente viable. Pero si lo fuera, ¿os parece que incluirlo como código es la forma correcta? ¿No sería más adecuado simplemente un fichero de texto que contenga información que posiblemente pueda ser variable para así no tener que editar el código cada vez que haya un cambio (o se implemente para un servicio diferente)?

mmnuria replied:
no entiendo a que se refiere con un fichero, ¿para las áreas?

JJ replied:
Aclaro

mmnuria replied:
Tienes razón, meter esta información directamente en el código no es buena idea porque hace que la lógica del programa quede demasiado ligada a los datos. Esto complica mucho el mantenimiento y hace que sea menos flexible para cambios o adaptaciones.

abbonno replied:
Areas ya no forma parte del proyecto, por lo que esta conversación queda desfasada y resuelta. Esto se debe a que no forma parte del modelo necesario para cumplir con el producto mínimo viable. [3087460]

JJ commented on internal/areas.go:
```diff
@@ -0,0 +1,18 @@
+package health_scheduler
+
+type Area struct {
```
En #3 ¿dice algo aparte de que existen áreas?

mmnuria replied:
Menciona que " Generalmente debe haber un enfermero por área, en total ocho. Cuando hay alguna baja o nos falla algún compañero hay sustitutos que lo cubren", por lo que hay que contemplar 1 de cada area y 2 sustitutos al dia.

JJ replied:
¿Y eso obliga a que crees un tipo?
Os recuerdo que modelar no es poner nombre a las cosas, y mucho menos poner un nombre inadecuado (llamar Area a sustitutos). Asegurar mínimos ¿te parece que sea parte del modelo o más bien de la lógica de negocio?

mmnuria replied:
Tienes razón, no sería necesario crear un tipo para eso, ya que pertenece a la logica de negocio

abbonno replied:
Ídem [3087460]

JJ commented on internal/areas.go:
```diff
@@ -0,0 +1,18 @@
+}
+
+var (
+	Intermedia1 = Area{Nombre: "Intermedia 1", EnfermerosNecesarios: 1}
```
¿Dónde dice en #3 que haya más un número determinado de enfermeros por área? Lo único que dice es que hay al menos uno por área.

mmnuria replied:
En la HU001 meciona que se necesita minimo uno por area y sustitutos que en el README se extiende esa explicación diciendo que se necesitan minimo 2 sustitutos tambien. Por eso mismo he contemplado en esta parte, que se necesita 1 enfermero por area y 2 sustitutos al dia, mínimo, por lo que, de esta forma lo dejo especificado como un requisito.

JJ replied:
Te recuerdo que tiene que ser un producto mínimo basado en lo que dice.

abbonno replied:
Ídem [3087460]

JJ reviewed (changes requested):
El concepto de producto mínimo implica que se cubra solo y exclusivamente lo que dice en la HU. Por otro lado, si las HUs son ambiguas o no tienen toda la información se puede inclurrir en muchos errores.
Finalmente, los mensajes de commit tienen que tener el formato:

```
#issue explicar la decisión (en 50 caracteres)
(línea en blanco)
Resto de la explicación
```

Eso permite a la persona que revisa ver más fácilmente de qué se está tratando.

JJ commented on internal/areas.go:
```diff
@@ -0,0 +1,18 @@
+	Urgencias2  = Area{Nombre: "Urgencias 2", EnfermerosNecesarios: 1}
+	Planta1     = Area{Nombre: "Planta 1", EnfermerosNecesarios: 1}
+	Planta2     = Area{Nombre: "Planta 2", EnfermerosNecesarios: 1}
+	Sustitutos  = Area{Nombre: "Sustitutos", EnfermerosNecesarios: 2}
```
¿"Sustitutos" es un área?

abbonno replied:
Ídem [3087460]

JJ reviewed (changes requested):
Por favor, considera los nuevos comentarios. Una vez más, @abbonno tienes que trabajar con esto, y si tienes código de más o de menos va a ser carga adicional para tu objetivo 4 que va a llegar en una fecha posiblemente inconveniente.

mmnuria committed:
Eliminación de Area, ya que, forma parte de la logica de negocio.

mmnuria commented:
@JJ ya lo he solucionado, como me ha sugerido, esa parte es más correcta en la logica de negocio. Además, meter esta información directamente en el código no es buena idea porque complica mucho el mantenimiento y hace que sea menos flexible para cambios o adaptaciones, es decir, poco escalable.

JJ reviewed (commented):
Cuando os digo que en el YAML tenéis que incluir una entidad, es porque es prácticamente imposible que esté un modelo completo sin una entidad en el sentido de DDD. Aquí no hay ninguna, y si no hay ninguna modelización de quién crea las asignaciones y es capaz de establecer las relaciones entre los diferentes objetos valor vas a tener que programarlos más adelante. @abbonno , antes de aprobar deberíais realmente pensar como vais a usarlo, porque una vez más, lo que no se quede hecho ahora, aprobado por vosotros mismos, vais a tener que hacerlo más adelante en un momento más inconveniente.
Esto por no mencionar que *realmente* debéis entender y poner en práctica el concepto de "modelo" usando la metodología DDD

mmnuria committed:
Soluciona #14 se ha decidido añadir un ID único para cada objeto

se ha añadido al struct que almacena la información de la asignación de un turno a un empleado un identificador unico "ID"

mmnuria committed:
Soluciona #14 se resuelve el problema de que no era una entidad

He tenido en cuenta que se va a generar cronogramas semanalmente

mmnuria committed:
Soluciona parte #14 se necesita asignar a cada empleado si es flexible o no para las bajas repentinas

Se ha eliminado datos redundantes

mmnuria commented:
@JJ solicito revision

JJ commented on internal/turno.go:
```diff
@@ -0,0 +1,31 @@
+
+	switch nombre {
+	case Mañana:
+		inicio = "08:00"
```
No se deben usar "números mágicos" y menos números magicos repetidos. Si hay un literal, aunque aparezca una sola vez, se debe usar una constante.

abbonno replied:
Finalmente se ha optado por no emplear este formato de hora.

JJ reviewed (commented):
Por favor, ten en cuenta las mejores prácticas.

mmnuria committed:
Eliminación archivos que he considerado innecesarios

mmnuria committed:
Soluciona #8 añado const para almacenar los valores que se repiten y creo los turnos para cada tipo especifico.

He reducido los archivos, ya que considero que eran innecesarios, además si el turno es nocturno tiene en cuenta que la hora de fin es al dia siguiente.

mmnuria committed:
Soluciona #13 la información de un empleado no es mas que su nombre y el tipo de contrato que tiene.

He modificado el struct para que cumpla SSUT.

mmnuria commented:
@JJ listo para revisión

JJ commented on internal/contrato.go:
```diff
@@ -0,0 +1,15 @@
+package health_scheduler
+
+type Contrato struct {
```
¿Esto tiene que ser parte del modelo o simplemente de la lógica de negocio? Supongo que ya que @abbonno lo ha aprobado tiene que ser parte del modelo. Si es así, ¿esto es una entidad o un objeto valor? Si la respuesta es la segunda, ¿desde dónde se maneja?

mmnuria replied:
Eso es cierto que estaba dudando si realmente se necesita ahora mismo, porque no lo menciona en la HU001, por lo que creo que lo voy a eliminar

abbonno replied:
Contrato eliminado al no formar parte del producto mínimo viable. [3f09dcb]

mmnuria committed:
Soluciona #14 guardamos en una lista los turnos que hay.

He tenido en cuenta que se crean los turnos y se almacenan en la lista por fechas y empleados asignados.

JJ commented on internal/empleado.go:
```diff
@@ -0,0 +1,13 @@
+
+type Empleado struct {
+	Nombre       string
+	TipoContrato Contrato
```
Si sólo se emplea `Contrato` desde `Empleado`, ¿por qué hay que crear un `struct` aparte?
Recordad que un modelo no es ponerle nombre a las cosas, sino ver claramente qué es cada uno y qué relaciones hay entre sí.
Y, una vez más, que un empleado tenga un nombre y un contrato *no implica* que los dos tengan que estar en la misma estructura de datos.

abbonno replied:
[8d3dbc6]

JJ reviewed (changes requested):
Los modelos tienen que tratar de ver la mejor forma de relacionar los diferentes datos. Domain Driven Design ayuda con eso, pero también pensar desde los dos lados, tú y @abbonno , cómo se va a usar. Si no repito lo mismo de siempre, os vais a encontrar en el objetivo 4 con problemas imprevistos por tener un modelo incompleto o incorrecto.

mmnuria committed:
Soluciona [https://github.com/abbonno/healthScheduler/pull/10#discussion_r1861999110](comentario) ya que no son parte del modelo del problema, además empleado.go no es necesario crearlo ya que no tiene relevancia.

mmnuria committed:
Soluciona #8 cambiamos el empleado por un string.

Solo guardará el nombre del empleado

mmnuria committed:
Soluciona #14 tiene en cuenta los cambios de empleado.go

mmnuria commented:
@JJ revision cuando puedas.

JJ commented on internal/cronograma.go:
```diff
@@ -0,0 +1,40 @@
+	e.Turnos = append(e.Turnos, turno)
+}
+
+func (e *RegistroTurnos) ListarTurnos() {
```
¿El modelo incluye alguna función?

mmnuria replied:
tiene razon, fallo mio. Lo estuvimos hablando en clase justamente.

abbonno replied:
[5d0a180]

JJ commented on internal/cronograma.go:
```diff
@@ -0,0 +1,40 @@
+	}
+}
+
+func (e *RegistroTurnos) AgregarTurno(turno Turno) {
```
No hay nada que diga más fuerte "he dejado que Copilot genere el código" que un montón de cosas que posiblemente no sean parte de este milestone.

mmnuria replied:
Es cierto, esto realmente no hace falta en el primer milestone.

abbonno replied:
[5d0a180]

JJ reviewed (changes requested):
Por favor, limítate a hacer lo que sea necesario en el milestone correspondiente. Y sí os rogaría que cuando hagáis algún cambio no pidáis revisión hasta que no esté aprobada.

mmnuria committed:
Soluciona #14 omisión de funciones que no son necesarias para el primer milestone.

Además el modelo no debe de incluir funciones.

mmnuria commented:
> Y sí os rogaría que cuando hagáis algún cambio no pidáis revisión hasta que no esté aprobada.

@JJ No entiendo bien a que se refiere con esto.

mmnuria committed:
Soluciona #14 he modificado el struct ya que no era necesario, guarda simplemente una lista.

abbonno commented on internal/turno.go:
```diff
@@ -0,0 +1,84 @@
+	}
+}
+
+func incrementarDia(fecha string) string {
```
No estoy muy seguro de si una función fuera del constructor es necesario para modelar la historia de usuario.
Editado: acabo de revisar el constructor y considero que es importante para no juntar código en el mismo para verificar que el día sea el siguiente, sigue las buenas prácticas.

JJ replied:
Hay que resolver todos los comentarios.

JJ replied:
Ya es dudoso que las funciones de validación sean parte del modelo. Si las funciones de validación necesitan una función auxiliar, realmente el modelo es excesivamente complicado y debería simplificarse *usando el sistema de tipos*

abbonno replied:
Se elimina la función incrementarDia, innecesaria. [75d9cc4]

abbonno replied:
[75d9cc4]

abbonno commented on internal/cronograma.go:
```diff
@@ -0,0 +1,8 @@
+
+type RegistroTurnos []Turno // Lista de turnos asignados
+
+func NuevoRegistroTurno() *RegistroTurnos {
```
Me he visto en un caso como este y no es necesario crear un constructor para este tipo, solamente en caso de que tengas que verificar que se cumple alguna condición (que no esté vacío el array turnos). Además, de la manera en que está escrito no recibe ningún argumento. Sugiero no crear el constructor.

abbonno replied:
[a9dd94c]

mmnuria committed:
#14 constructor no necesario.

mmnuria commented:
muchas gracias por tus correcciones @abbonno , solicito revision @JJ

JJ commented:
> > Y sí os rogaría que cuando hagáis algún cambio no pidáis revisión hasta que no esté aprobada.
> 
> @JJ No entiendo bien a que se refiere con esto.

Estás trabajando en el repositorio de @abbonno, y él tiene que trabajar con este código. Hay que esperar a que él lo revise y eventualmente aprueba para hacerlo yo.

JJ commented on internal/cronograma.go:
```diff
@@ -0,0 +1,3 @@
+package health_scheduler
+
+type RegistroTurnos []Turno // Lista de turnos asignados
```
1. ¿Que el nombre del fichero no tenga nada que ver con lo que contiene sigue las mejores prácticas?
2. Si el comentario no coincide con el nombre del tipo, ¿cuál de los dos está mal? Respuesta: los dos
3. ¿Te parece buena modularización crear un fichero solo para definir un tipo?

abbonno replied:
Tienes razón, lo que fue pensado como un cronograma con más funciones ahora es una lista de turnos. Debería quizás ir dentro del módulo turno.go. Después de esto (si lo haces así) recuerda cambiar la entidad en el iv.yaml.

abbonno replied:
[f520cdd]

JJ commented on internal/turno.go:
```diff
@@ -0,0 +1,84 @@
+type HorarioMañana string
+
+const (
+	InicioMañana HorarioMañana = "08:00"
```
@abbonno ¿te parece esta una forma sencilla de trabajar con los inicios de turno? ¿Qué código tendrías que hacer para sacar el inicio del turno?

abbonno replied:
Al principio no estaba seguro de esta implementación por eso mismo, sin embargo esto puede facilitar la legibilidad del código en el futuro. Pongamos el siguiente caso, se quiere contar el tiempo de la jornada de un enfermero.

1. **Caso uno:** pongamos que solo disponemos del dato TipoTurno, sabemos que tiene turno de mañanas. Para saber el tiempo que ha estado trabajando tendríamos que añadir un atributo "duraciónTurno" ya que esta difiere (en contreto de mañana o tarde al de noche). Sin embargo de este modo lo estaríamos definiendo un modelo muy concreto y no podrá ser muy útil para futuros casos en la lógica de negocio.
2. **Caso dos:** en la misma situación de antes, en vez de usar otro atributo de duración del turno, simplemente calculamos el tiempo del turno almacenado y el siguiente turno. ¿Cómo sabemos cuál es el siguiente turno? Habría que escribir código extra, además de que para referirnos a un turno, imaginemos, de mañana, tendremos que usar información del horario de tardes y viceversa.

El último caso (o casos), y el que hace, a mi parecer, de esta implementación la adecuada es: tenemos un enfermero de mañanas y para especificar su tiempo trabajando hacemos lo mismo que en el caso 1 (calculamos a partir del inicio) pero empleando un fin de jornada. Difiere del caso 2 en que para calcularlo no estamos diciendo "desde cuando empieza un turno hasta que empieza el siguiente", sino desde cuando empieza a cuando acaba. A esto se le añade que en otros casos podamos utilizar este "fin de jornada" en vez de escribir código extra para obtener cuál es la jordana siguiente y sacar su inicio.
En fin, he intentado ser lo más claro posible indicando las posibles facilidades que, a mi parecer, aporta esta estructura y que define bien el modelo que se quiere representar, a pesar de requerir algunas líneas más de código.

JJ replied:
> Al principio no estaba seguro de esta implementación por eso mismo, sin embargo esto puede facilitar la legibilidad del código en el futuro. Pongamos el siguiente caso, se quiere contar el tiempo de la jornada de un enfermero.

¿Qué os tengo dicho sobre el código que se hace para no sé qué "en un futuro"?

abbonno replied:
Se simplifica la estructura de Turno. [75d9cc4]

JJ commented on internal/turno.go:
```diff
@@ -0,0 +1,84 @@
+	var inicio, fin, fechaFin string
+
+	switch nombre {
+	case Mañana:
```
... este es el código que tienes que hacer, precisamente. 15 líneas de código ¿te parece que indiquen que se ha hecho correctamente la modelización?

abbonno replied:
Absolutamente. Estas 15 líneas son necesarias, incluso ínfimas, dado que más tarde en otros apartados obtendremos esta información con tan solo referirnos al string "mañana", "tarde" o "noche". Aporta, a mi parecer, información necesaria y crean un modelo inmutable y fácil de emplear (dadas, además, las indicaciones en el comentario anterior sobre esto mismo).

JJ replied:
¿De veras no se te ocurre ninguna otra alternativa?

abbonno replied:
Claro que se me ocurre, y aunque no se me ocurriera, como bien decimos en clase, siempre hay varias formas de hacer las cosas, aunque no sean las mejores. Aquí específicamente, entiendo que el acercamiento de mi compañera no es el mejor, pero tiene sus ventajas, por eso lo he considerado correcto.

JJ replied:
El mejor código es el que no existe. Si puedes evitar 15 líneas de código con una estructura de datos, hazlo.

abbonno replied:
De acuerdo

abbonno replied:
Código farragoso modificado. [d6a66dc]

JJ commented on internal/turno.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+type Turno struct {
+	Empleado    string
```
Si el nombre del empleado está *dentro* del turno
1. ¿Cómo garantizas que no le estás asignando a un empleado dos turnos diferentes?
2. ¿Cómo accedes al turno que se le ha asignado a cada empleado?

abbonno replied:
1. A un empleado se le **van** a asignar dos turnos diferentes. Dos, tres y los que sea. El registro de turnos solo contiene quién va a estar en un área a qué hora. Desde que empieza a trabajar, el enfermero tiene definidos sus horarios de forma estática (aunque después se pueda intercambiar con otros enfermeros) y dependiendo de su tipo de contrato podrá tener turnos de mañana, de tardes y de noches.
2. Para este caso sí que veo necesario crear una entidad enfermero que guarde sus turnos, así como otra información sobre su contrato.

JJ replied:
Perdona, @abbonno , pero los comentarios se resuelven o cambiando el código o mediante cambios en otro lado, y por parte de @mmnuria. Si respondes, soy yo el que tengo que decidir si está resuelto o no, que soy el que ha planteado el comentario.
Y no, no está resuelto con esa pregunta ni mucho menos. Ni tenías tú que responderla, por cierto.

abbonno replied:
Se ha provisto del siguiente acercamiento a la definición de un conjunto de turnos que le pertenecen únicamente a un empleado. [40d702f]

JJ replied:
¿has leído donde pone "por parte de @mmnuria"? ¿Estás justificando tú donde se ha resuelto algo en vez de hacerlo ella?

JJ reviewed (changes requested):
El principal problema es que hay que resolver todos los comentarios. El resto también tienen su importancia. Con el código nuevo que se genere, se aconseja que se consulten de nuevo los errores frecuentes.

mmnuria committed:
Soluciona [https://github.com/abbonno/healthScheduler/pull/10#discussion_r1865402014](comentario) eliminación del archivo innecesario

mmnuria committed:
Soluciona [https://github.com/abbonno/healthScheduler/pull/10#discussion_r1866490914](comentario) para reducir el codigo y añade RegistroTurnos

mmnuria committed:
Soluciona [https://github.com/abbonno/healthScheduler/pull/10#discussion_r1865405474](comentario) creacion de la entidad de empleado para almacenar información necesaria.

mmnuria commented:
@abbonno cuando puedas me dices que opinas de los cambios.

abbonno commented on internal/empleado.go:
```diff
@@ -0,0 +1,17 @@
+type Empleado struct {
+	ID           string  // Identificador único (puede ser un UUID o similar)
+	Nombre       string  // Nombre del enfermero
+	TipoContrato string  // Tipo de contrato (puede ser "Tiempo completo", "Medio tiempo", etc.)
```
Deberías añadir una restricción para que el tipo contrato sea uno de los posibles (los de la documentación).

JJ replied:
En general, en PRs sólo se hacen preguntas sobre qué se ha considerado o qué no, salvo que sean errores flagrantes, en cuyo caso se puede hacer una sugerencia directamente sobre el código (icono con un +-)

mmnuria replied:
pero en teoria la parte del tipo de contrato no es necesaria para el modelo del milestone 0. @abbonno

abbonno replied:
lo reformulo entonces, crees que el modelo de la hu 1 (en concreto el tema de los turnos) está completo si no se concretan qué tipos de turnos llevan a cabo los empleados (en este caso, de su tipo de contrato)?

mmnuria replied:
podría ser útil contemplar que cada turno tiene que asociarse a un empleado teniendo en cuenta su contrato. Eso lo hice al inicio, pero al no formar parte del milestone 0 y HU001 lo eliminé. Si lo modificas puedo añadirlo sin problema. @abbonno

abbonno replied:
@JJ es posible modificar la historia de usuario a estas alturas? o es preferible asumir que al no venir directamente implícito es innecesario hacer esto que he planteado?

mmnuria replied:
En principio según tratamos en clase de prácticas, llegamos a la conclusión de que por el momento no haría falta centrarse en los contratos ya que no es algo que englobe este milestone. @JJ esperamos que nos digas para poder continuar y superar el objetivo cuanto antes. Gracias

JJ replied:
@abbonno a estas alturas mejor que no lo hagas. Y efectivamente, no parece que los contratos sean importantes.

abbonno replied:
entonces queda todo aprobado hasta ahora

JJ replied:
¿Si? ¿Para qué vas a usar esto en la lógica de negocio?

abbonno replied:
Se opta finalmente por no recurrir a TipoContrato. [75d9cc4]

mmnuria commented:
@JJ solicito revision

JJ commented:
Hay una opción de "Review" específica para aprobar. Por favor, usadla.

JJ commented on internal/empleado.go:
```diff
@@ -0,0 +1,17 @@
+package health_scheduler
+
+type Empleado struct {
+	ID           string  // Identificador único (puede ser un UUID o similar)
```
¿Esto es un objeto valor? Si lo es, ¿para qué necesita el identificador único?

abbonno replied:
Cambio en la estructura de Empleado [75d9cc4]

JJ commented on internal/empleado.go:
```diff
@@ -0,0 +1,17 @@
+
+type Empleado struct {
+	ID           string  // Identificador único (puede ser un UUID o similar)
+	Nombre       string  // Nombre del enfermero
```
¿El comentario aporta algo? Si lo hace, ¿por qué no lo pones en el identificador?
Los comentarios inline deben evitarse siguiendo la práctica de código limpio.

abbonno replied:
No lo aportan, siguiendo buenas prácticas se elimina. [75d9cc4]

JJ commented on internal/empleado.go:
No acabo de entender qué utilidad tiene esto para asignar turnos. Y si la tiene, ¿qué utilidad puede tener el tipo de contrato?

mmnuria replied:
Segun me dijo mi compañero @abbonno le parecia mas interesante tener una entidad de empleados con esta información. Yo tampcoo creo que tenga una gran utilidad ya que, en turno.go ya tenemos esa información sin necesidad de incorporar otra entidad aparte (empleado)

JJ replied:
Ni este comentario ni el siguiente se han resuelto correctamente.

abbonno replied:
La utilidad para asignar turnos es que cada empleado del tipo Empleado tiene asignado un vector de Turnos en TurnoAsignado, que se refiere a estos (los empleados) mediante un puntero ya que es necesario referirse a ellos directamente. ¿Qué ocurre si el empleado se cambia de nombre? 
Lo primero creo que ya estaba resuelto una vez modificado por mi compañera, pero dejo explicado el cambio y resuelvo.

JJ commented on internal/empleado.go:
```diff
@@ -0,0 +1,17 @@
+		ID:           id,
+		Nombre:       nombre,
+		TipoContrato: tipoContrato,
+		Turnos:       []Turno{}, // Inicialmente sin turnos asignados
```
Lo esencial de la modelización es la gestión de la mutabilidad. Si pones todo junto en un solo objeto todo lo que corresponde al mismo, es que no estás entendiendo este concepto. DDD consiste precisamente en organizar el código en objetos valor y entidades y en relacionar los mismos, comprendiendo y entendiendo cómo se maneja esa mutabilidad. Una vez más, DDD no es asignar nombres a las cosas.

abbonno replied:
Se opta finalmente por cambiar la estructura de turno [75d9cc4].
Más específicamente el array de turnos correspondiente a cada empleado se saca del struct empleado para dar lugar a una nueva entidad propia.

JJ replied:
No entiendo. ¿Estás diciendo lo que está haciendo @mmnuria ? ¿Y en sitios donde ya se había resuelto algo porque se había cambiado el código?

JJ reviewed (changes requested):
No se trata solo de "meter entidades" o de "convertir un objeto valor en una entidad metiéndole un ID". Se trata de entender cómo se gestiona la mutabilidad del modelo y desde dónde hay que hacerlo.
@abbonno, sería conveniente también qu etú revisaras este tipo de cosas. Ya has superado el objetivo 2, deberías tener estos conceptos bien claros. Pero sobre todo deberías de tener en cuenta lo que vas a usar y como. Si no haces un mínimo esfuerzo ahora de ver cómo vas a usar este modelo antes de llegar al objetivo 4, tendrás que hacer triple esfuerzo para usarlo cuando efectivamente llegues. Con esto te quiero decir que este PR no es el problema de @mmnuria ni mi problema, es *tu* problema y como tal deberías tratarlo.

mmnuria committed:
Soluciona #13 se modifica campo empleado a string

mmnuria commented:
Segun lo que me ha mencionado @JJ y tambien mi opinión sobre como resolver este objetivo, he dejado un nuevo commit con los cambios de archivos que he realizado.

Considero que con tener una entidad turno como la que se puede ver en el codigo es mas que suficiente para abarcar el milestone 0. Esto es, ya que, la entidad de empleado (por ahora) no es necesaria para el modelo del milestone 0, ya que, lo unico que guardaría sería su nombre y para eso no es necesario crear una entidad especifica.

@abbonno y @JJ decirme cual es vuestra opinion al respecto y si considerais que ya he superado el objetivo.

JJ commented:
El commit [c1da3a7](https://github.com/abbonno/healthScheduler/pull/10/commits/c1da3a72ad2e96bf0469ab7856ec4f6f87e52139) no tiene ninguna referencia a issue. En general, eso conduce a código erróneo y en todo caso complicado o imposible de evaluar. Procedo en todo caso a revisarlo.

JJ commented on internal/turno.go:
```diff
@@ -0,0 +1,60 @@
+	Fin         string
+}
+
+var RegistroTurnos []Turno // Lista de turnos asignados
```
¿De qué periodo es este registro de turnos?

JJ commented on internal/turno.go:
¿La estructura de datos más importante aquí es `Turno` o `RegistroTurnos`? En función de cuál lo sea, ¿el nombre del fichero sigue las mejores prácticas?

JJ replied:
Hay que resolver correctamente *todos* los comentarios.

JJ reviewed (changes requested):
Aparte de no estar resueltos los comentarios, (ni resueltos por la persona adecuada), tendría que revisar el PR @abbonno antes de aprobarlo yo.

mmnuria committed:
Soluciona #8 reestructuración del turno

Se ha simplificado la entidad Turno para que contenga el mínimo de información necesaria. Además se ha usado un tipo más apropiado para el campo Fecha y se ha eliminado el campo Empleado de la entidad.

mmnuria committed:
Soluciona #13 relaciona empleado con sus turnos

Se establece la relación entre un empleado y sus turnos, de forma que se cumplan los requisitos de HU001

mmnuria committed:
Soluciona #13 cambio entidad principal

mmnuria commented:
Me he dado cuenta de que el enfoque que le estaba dando no era el más adecuado y tras revisarlo cuidadosamente, he realizado una serie de cambios que tratan de resolver el problema de una forma mínima más sencilla, permitiendo a cada empleado saber cuáles son sus turnos.

@abbonno solicito revisión.

abbonno commented on internal/empleado.go:
```diff
@@ -0,0 +1,5 @@
+
+type Empleado string
+
+type TurnosAsignados map[*Empleado][]Turno
```
Turnos asignados no debería ser donde se guarden los turnos?

abbonno replied:
Seguía pensando que debería ser una variable privada, pero a la vez me crea la duda el hecho de que se puedan definir métodos para actuar con los turnos. Una alternativa que he estado barajando sería hacer un singleton, con el que podríamos crear funciones específicas para controlar los turnos (obtener, cambiar, ingresar otro turno...), mientras que nos aseguramos de que solamente haya una instancia de TurnosAsignados. Al mismo tiempo, creo que esto es demasiado restrictivo para lo que queremos modelar y dificulta el desarrollo. Es por eso que, finalmente, concluyo en que sí que es necesario que TurnosAsignados sea un tipo y se cree una variable del tipo para almacenar los turnos de cada empleado.

JJ replied:
@abbonno es @mmnuria quien está haciendo este trabajo. Si responde correctamente a las historias de usuario y los problemas están bien planteados, lo que tú pienses no puede tener más relevancia que lo que ella piense, y más si no lo estás justificando en base a ninguno de los issues.

mmnuria committed:
Soluciona #8 variable donde se guardan los turnos

mmnuria commented:
@abbonno arreglado

abbonno commented:
> @abbonno arreglado

Correcto. Pues hacemos repaso de todo lo último: quedamos en que no es necesario representar el tipo de contrato, ni el área; turno contiene la información exclusivamente necesaria para modelar bien el Turno, así como Empleado lo tiene del empleado y por último existe una variable TurnosAsignados que contiene toda la información y relaciona a los empleados con sus turnos.
Por mi parte tiene el visto bueno.

mmnuria commented:
@JJ cuando puedas necesito revisión.

JJ commented on internal/empleado.go:
¿De veras que necesitáis un fichero para dos definiciones?

mmnuria replied:
Hay que tener en cuenta que, más adelante cuando @abbonno necesite modelar la parte de los contratos y las horas que cada empleado tiene establecido por contrato, necesitará este fichero para incluir esa informacion al empleado y cambiarlo a type struct, por lo que, considero que es una forma flexible para ello.

JJ replied:
Necesitará *un* fichero donde esté incluido.

abbonno replied:
No comprendo bien a qué te refieres con indicar que se necesita un fichero cuando el problema que estás planteando es precisamente que hemos creado un fichero para esas dos definiciones. ¿Podrías ser un poco más preciso?

JJ replied:
> Hay que tener en cuenta que, más adelante cuando @abbonno necesite modelar la parte de los contratos y las horas que cada empleado tiene establecido por contrato, necesitará este fichero para incluir esa informacion al empleado y cambiarlo a type struct, por lo que, considero que es una forma flexible para ello.

Tú dices "necesitará *este* fichero". Yo digo "necesitará *un* fichero". Por favor, leed también vuestras respuestas aparte de las mías. El contexto está todo aquí en 10 líneas.

JJ replied:
En este commit, además c1da3a72ad2e96bf0469ab7856ec4f6f87e52139 no hay referencia a ningún issue ni HU, por lo tanto en todo caso el código no se puede evaluar al no seguir las buenas prácticas.

JJ replied:
Esto sigue sin solucionar.

JJ commented on internal/empleado.go:
```diff
@@ -0,0 +1,5 @@
+
+type Empleado string
+
+var TurnosAsignados map[*Empleado][]Turno
```
... una de las cuales es una variable global *que no debe usarse nunca*. Aunque tenga ámbito de paquete.

mmnuria replied:
Como puede ver en [el comentario](https://github.com/abbonno/healthScheduler/pull/10#discussion_r1871243016), @abbonno me sugirió que fuese una variable, pero es cierto que, es más apropiado que sea un `type`

JJ replied:
Y tú seguiste un comentario que no tiene nada que ver con los issues ni con las historias de usuario, creando trabajo extra para todos, incluyéndome a mi. Repetidamente os he dicho que los comentarios no son para sugerencias, sino que sirven para plantear alternativas o entender mejor qué es lo que se ha planteado. Igual que una HU no es un mandao, un milestone no es un mandao, un comentario a un PR *tampoco es un mandao*, y esto tenéis que entenderlo los dos.

abbonno replied:
Finalmente se opta por modelar TurnosAsignados como una entidad [bde5ecf]

JJ reviewed (changes requested):
El principal problema es el uso de la variable global. Aparte de la mala práctica que implica el uso de la misma, una variable *no* es un modelo.
 Tampoco sé si quitarse de enmedio las horas máximas que puede estar un empleado es razonable, pero @abbonno lo ha aprobado, así que él sabrá.

mmnuria commented:
> El principal problema es el uso de la variable global. Aparte de la mala práctica que implica el uso de la misma, una variable _no_ es un modelo. Tampoco sé si quitarse de enmedio las horas máximas que puede estar un empleado es razonable, pero @abbonno lo ha aprobado, así que él sabrá.

Con respecto a las horas máximas de un empleado no lo he añadido ya que, al centrarnos en el MPV de la HU001, de momento no sería necesario incluirlo siendo estrictos con sus recomendaciones. Por lo que, me he ceñido exclusivamente a modelar ese MPV que resuelve el milestone 0.

Lo estrictamente mínimo necesario (segun HU001) es poder guardar la informacion de cada enfermero con los turnos que se le han asociado para que no moleste al resto de personal, pudiendo acceder a traves de la aplicación para que sepa que turno le toca (HU001)

mmnuria committed:
Soluciona #13 los turnos asociados a un empleado como type

mmnuria commented:
@abbonno he realizado nuevos cambios. Espero tu respuesta.

abbonno reviewed (changes requested):
Creo que a lo que se está refiriendo el profesor es al hecho de que la variable sea global. Es una mala práxis: reduce la modularidad y el fácil entendimiento del código, ya que al ser empleada en muchos ámbitos distintos es difícil seguirle el rastro en todo momento durante el desarrollo. En go es posible restringir el uso de una variable a solo al paquete comenzándola con minúsculas.

mmnuria commented:
> Creo que a lo que se está refiriendo el profesor es al hecho de que la variable sea global. Es una mala práxis: reduce la modularidad y el fácil entendimiento del código, ya que al ser empleada en muchos ámbitos distintos es difícil seguirle el rastro en todo momento durante el desarrollo. En go es posible restringir el uso de una variable a solo al paquete comenzándola con minúsculas.

El propósito de este `type` es definir la estructura que se usará para tipar futuras variables que serán manejadas en el programa, dentro de funciones o métodos específicos según corresponda. Crear una variable directamente aquí, ya sea global o local, no tendría sentido, ya que la intención es establecer un esquema para datos futuros, o estoy entendiendo algo mal @JJ ?

JJ commented:
> > El principal problema es el uso de la variable global. Aparte de la mala práctica que implica el uso de la misma, una variable _no_ es un modelo. Tampoco sé si quitarse de enmedio las horas máximas que puede estar un empleado es razonable, pero @abbonno lo ha aprobado, así que él sabrá.
> 
> Con respecto a las horas máximas de un empleado no lo he añadido ya que, al centrarnos en el MPV de la HU001 me he ceñido exclusivamente a modelar ese MPV que resuelve el milestone 0.
> 
> Lo estrictamente mínimo necesario (segun HU001) es poder guardar la informacion de cada enfermero con los turnos que se le han asociado para que no moleste al resto de personal, pudiendo acceder a traves de la aplicación para que sepa que turno le toca (HU001)

No te pedía que contestaras a esto. Ya he dicho que @abbonno lo ha aprobado, así que él sabrá. Pero ya que has contestado, ¿os parece (a los dos) que no saber cuantas horas totales hay que asignar, sea total, parcial o mediopensionista, es aceptable? Mientras que las *variables* no son parte del modelo, las *constantes* sí que lo son.

JJ reviewed (commented):
Añado solo como comentario, ya que el propietario del repositorio todavía no ha aprobado. No es una buena práctica discutirle al propietario del repositorio aludiendo al profesor (o persona más senior en general) esperando que le corrija. La persona que lleva el proyecto es quien tiene que vivir con esos cambios, así que sería conveniente que se siguiera sus peticiones de cambios. Si hay alguna duda por ambas partes, es mejor que la resolváis primero entre vosotros y luego si no se entiende algo, podéis acudir a mi.

abbonno commented:
> > > El principal problema es el uso de la variable global. Aparte de la mala práctica que implica el uso de la misma, una variable _no_ es un modelo. Tampoco sé si quitarse de enmedio las horas máximas que puede estar un empleado es razonable, pero @abbonno lo ha aprobado, así que él sabrá.
> > 
> > 
> > Con respecto a las horas máximas de un empleado no lo he añadido ya que, al centrarnos en el MPV de la HU001 me he ceñido exclusivamente a modelar ese MPV que resuelve el milestone 0.
> > Lo estrictamente mínimo necesario (segun HU001) es poder guardar la informacion de cada enfermero con los turnos que se le han asociado para que no moleste al resto de personal, pudiendo acceder a traves de la aplicación para que sepa que turno le toca (HU001)
> 
> No te pedía que contestaras a esto. Ya he dicho que @abbonno lo ha aprobado, así que él sabrá. Pero ya que has contestado, ¿os parece (a los dos) que no saber cuantas horas totales hay que asignar, sea total, parcial o mediopensionista, es aceptable? Mientras que las _variables_ no son parte del modelo, las _constantes_ sí que lo son.

Eso se hará una vez sea necesario modelar el tipo de contrato. Mientras tanto, al no haber una clasificación tal, no podemos asignar unas horas a cada usuario.

mmnuria commented:
Solicito revision @JJ

JJ reviewed (changes requested):
Los comentarios se deben resolver con un cambio o indicarse qué se ha cambiado. Es inaceptable que se cierren simplemente, y @abbonno no debería haber aceptado el PR sin haber comprobado previamente que se ha hecho así.

JJ reviewed (changes requested):
Aparte de lo indicado, no puede haber código que no referencie un issue.

mmnuria commented:
@JJ ya he realizado un rebase del commit que no seguía las buenas prácticas.

JJ reviewed (changes requested):
Seguís sin resolver los comentarios.

mmnuria commented:
@JJ comentarios resueltos.

JJ commented:
¿De qué forma darle al botón "resolve" es "resolver el comentario"? ¿Qué habéis resuelto o cambiado?

JJ commented on internal/empleado.go:
```diff
@@ -0,0 +1,5 @@
+package health_scheduler
+
+type Empleado string
```
Os recuerdo también que habéis hecho esto tratando de resolver #13, es decir, respondiendo a la pregunta ¿qué información hay que tener sobre un empleado para asignarle un turno específico? Lo que habéis hecho aquí no tiene que ver más que vagamente con ese issue.

mmnuria replied:
el issue fue creado inicialmente para tratar diferentes partes en las que se dividía el empleado (nombre, tipo de contrato, etc) las cuales finalmente se han ido modificando y llegando a la conclusion de que no tienen relevancia para este objetivo. A la hora de hacer el commit, referencié a ese issue para que se pudiera tener en cuenta esos cambios que han ido surgiendo con respecto al empleado.

JJ reviewed (changes requested):
Y además, tened en cuenta esto. En realidad no tiene más solución que el que expliquéis en el mensaje del commit de qué forma estáis solucionando ese problema. Y de camino el comentario que no habéis resuelto.

mmnuria committed:
Soluciona [comentario](https://github.com/abbonno/healthScheduler/pull/10/files#r1872971098) acerca del uso de un fichero adicional solo para dos definiciones.

mmnuria committed:
Soluciona [comentario](https://github.com/abbonno/healthScheduler/pull/10/files#r1872971098) se añade las definiciones al archivo turno.go

mmnuria committed:
Soluciona [comentario](https://github.com/abbonno/healthScheduler/pull/10/files#r1872971098) modifica la entidad.

JJ reviewed (commented):
Por favor, resolved todos los comentarios.

mmnuria committed:
Soluciona #9 se determina con un struct el Turno y mediante TurnosAsignados se relaciona empleado con sus turnos.

Siguiendo las buenas prácticas.

mmnuria committed:
Soluciona #9 estableciendolo como entidad.

mmnuria commented:
@JJ cambios realizados.

mmnuria committed:
soluciona el seguimiento de las buenas prácticas

mmnuria committed:
Soluciona #13 nombre nuevo de la entidad

JJ reviewed (approved):
👍
